#!/usr/bin/env python3
PACKAGE = "ltu_actor_route_yolov8_detector"  # Catkin Package Name

# Import library
from dynamic_reconfigure.parameter_generator_catkin import *

# Create Instance
gen = ParameterGenerator()

gen.add('vest_mask_l_hue',       double_t,   0, 'vest_mask_l_hue',         15  , 0, 180)
gen.add('vest_mask_h_hue',       double_t,   0, 'vest_mask_h_hue',         30, 0, 180)
gen.add('vest_mask_l_sat',       double_t,   0, 'vest_mask_l_sat',         80  , 0, 255)
gen.add('vest_mask_h_sat',       double_t,   0, 'vest_mask_h_sat',         255, 0, 255)
gen.add('vest_mask_l_lum',       double_t,   0, 'vest_mask_l_lum',         90  , 0, 255)
gen.add('vest_mask_h_lum',       double_t,   0, 'vest_mask_h_lum',         255, 0, 255)

gen.add(
    "debug",  # variable name
    bool_t,  # data type
    0,  # msg level
    ">>> Publishes images to associated debug topics",  # display on hover
    True,  # default
)

gen.add(
    "enable_vest_mask",  # variable name
    bool_t,  # data type
    0,  # msg level
    ">>> Enables display of vest mask",  # display on hover
    False,  # default
)

gen.add(
    "enable_sign_box",  # variable name
    bool_t,  # data type
    0,  # msg level
    ">>> Enables display of vest mask",  # display on hover
    True,  # default
)

gen.add(
    "enable",  # variable name
    bool_t,  # data type
    0,  # msg level
    ">>> Pauses detection on uncheck but keeps the node running",  # display on hover
    True,  # default
)

gen.add(
    "flip_image",  # variable name
    bool_t,  # data type
    0,  # msg level
    ">>> Flips camera image on both axis",  # display on hover
    False,  # default
)

gen.add(
    "image_resize",  # variable name
    int_t,  # data type
    0,  # msg level
    ">>> Resizes camera image for detection",  # display on hover
    1080,  # default size
    192,  # Minimum size
    2560,  # Maximum size
)

gen.add(
    "pothole_lane_change",  # variable name
    bool_t,  # data type
    0,  # msg level
    ">>> Enables lane change pothole detection",  # display on hover
    False,  # default
)

gen.add(
    "pothole_image_resize",  # variable name
    int_t,  # data type
    0,  # msg level
    ">>> Resizes pothole image for detection",  # display on hover
    1080,  # default size
    192,  # Minimum size
    2560,  # Maximum size
)

gen.add(
    "pothole_crop_top",  # variable name
    int_t,  # data type
    0,  # msg level
    ">>> Adjusts pothole crop top position",  # display on hover
    0,  # default size
    192,  # Minimum size
    2560,  # Maximum size
)

gen.add(
    "pothole_crop_bottom",  # variable name
    int_t,  # data type
    0,  # msg level
    ">>> Adjusts pothole crop bottom position",  # display on hover
    0,  # default size
    192,  # Minimum size
    2560,  # Maximum size
)

gen.add(
    "pothole_crop_left",  # variable name
    int_t,  # data type
    0,  # msg level
    ">>> Adjusts pothole crop left position",  # display on hover
    0,  # default size
    192,  # Minimum size
    2560,  # Maximum size
)

gen.add(
    "pothole_crop_right",  # variable name
    int_t,  # data type
    0,  # msg level
    ">>> Adjusts pothole crop right position",  # display on hover
    0,  # default size
    192,  # Minimum size
    2560,  # Maximum size
)

# Generate the menu in rqt tool
exit(gen.generate(PACKAGE, "route_yolov8_detector", "YoloDetect"))
# 1st param - Package name constant from top
# 2nd param - node name this could be called in (for documentation only)
# 3rd param - This file's name without extension. Adds 'Config' at the end

# for details: https://wiki.ros.org/dynamic_reconfigure/Tutorials/HowToWriteYourFirstCfgFile
